nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/handler/SMTPMessageHandlerFactory.java

```diff
@@ -0,0 +1,147 @@
+        public void recipient(String recipient) throws RejectException {
+            // TODO: possibly whitelist receivers?
+            this.recipient = recipient;
+        }
+
+        @Override
+        public void data(InputStream inputStream) throws RejectException, TooMuchDataException, IOException {
+            // Start counting the timer...
+
+            StopWatch watch = new StopWatch(false);
+
+            SMTPServer server = messageContext.getSMTPServer();
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            byte [] buffer = new byte[1024];
+            int rd;
+
+            while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1) {
+                baos.write(buffer, 0, rd);
+            }
+            if (baos.getBufferLength() > server.getMaxMessageSize()) {
+                throw new TooMuchDataException("Data exceeds the amount allowed.");
+            }
+
+            baos.flush();
+            this.messageBody = baos.toByteArray();
+
+
+            X509Certificate[] certificates = new X509Certificate[]{};
+
+            String remoteIP = messageContext.getRemoteAddress().toString();
+            String helo = messageContext.getHelo();
+            if (messageContext.getTlsPeerCertificates() != null ){
+                certificates = (X509Certificate[]) messageContext.getTlsPeerCertificates().clone();
+            }
+
+            SmtpEvent message = new SmtpEvent(remoteIP, helo, from, recipient, certificates, messageBody);
+            try {
+                    // Try to queue the message back to the NiFi session
+                    incomingMessages.put(message);
+            } catch (InterruptedException e) {
+                // Throws an error to the  incoming server alerting about issues queuing the message in NiFi
+                logger.error("Hit an error sending message back to NiFi main thread {}. Sending and error back to SMTP client and discarding message", new Object [] {e});
+                throw new DropConnectionException(451, "NiFi something went wrong while processing you message, please retry again later");
+            }
+
+            // Once message has been sent to the queue, it should be processed by NiFi onTrigger,
+            // a flowfile created and its processed status updated before an acknowledgment is
+            // given back to the SMTP client
+            try {
+                synchronized(message) {
+                    while(! message.getProcessed()) {
+                        // Check to see if it is too late...
+                        final long serverTimeout = TimeUnit.NANOSECONDS.convert(messageContext.getSMTPServer().getConnectionTimeout(), TimeUnit.MILLISECONDS);
+                        if ( watch.getElapsed(TimeUnit.MILLISECONDS) <=  serverTimeout) {
+                            // Will should allow for a couple of retries;
+                            message.wait(messageContext.getSMTPServer().getConnectionTimeout() / 20);
```

Comment:
@ijokarumawak you are correct. The move to done tries to force the handler to wait for an downstream acknowledgement before ack'ing back to the sending client.

Can you explain where how are you hitting the bug?

I tried reproducing with

Max sessions = 1

connect to port
```
220 bermudatriangle ESMTP Apache NiFi
helo .
250 bermudatriangle
mail from: a
250 Ok
rcpt to: b
250 Ok
data
354 End data with <CR><LF>.<CR><LF>
aaaa
.
250 Ok

```

Meanwhile other host gets:

```
$ telnet 0 1025
Trying 0.0.0.0...
Connected to 0.
Escape character is '^]'.
421 Too many connections, try again later
Connection closed by foreign host.

```

If I set Maximum to 2:

```
$ telnet 0 1025
Trying 0.0.0.0...
Connected to 0.
Escape character is '^]'.
220 bermudatriangle ESMTP Apache NiFi
helo .
250 bermudatriangle
mail from: x
250 Ok
rcpt to: 2
250 Ok
data
354 End data with <CR><LF>.<CR><LF>
aaaa
.
250 Ok

```

```
$ telnet 0 1025
Trying 0.0.0.0...
telnet: connect to address 0.0.0.0: Connection refused
[amiranda@development ~]$ telnet 0 1025
Trying 0.0.0.0...
Connected to 0.
Escape character is '^]'.
220 bermudatriangle ESMTP Apache NiFi
helo .
250 bermudatriangle
mail from: x
250 Ok
rcpt to: 1
250 Ok
data
354 End data with <CR><LF>.<CR><LF>
bbbb
.
250 Ok

```